analyzer/src/types.rs

```diff
@@ -17,97 +21,159 @@ pub struct Typing {
-        let first = types.next().unwrap_or(NOTHING);
-        types.try_fold(first, |acc, ty| {
-            self.convert_description(acc, ty)
-                .or_else(|_| self.convert_description(ty, acc))
-        })
+    if rvalue == assign_to {
+        // if both references are the same then no need to lookup, it's the same type
+        return Ok(assign_to);
     }
 
-    pub(crate) fn add_type(&mut self, ty: Type) -> TypeId {
-        let type_id = TypeId(self.types.len());
-        self.types.push(ty);
-        type_id
+    let lhs = get_type(assign_to, ura).unwrap_or_else(|| panic!("cannot find type {assign_to:?}`"));
+    let rhs = get_type(rvalue, ura).unwrap_or_else(|| panic!("cannot find type {rvalue:?}`"));
+    if lhs == rhs {
+        return Ok(assign_to);
     }
 
-    pub(crate) fn get_type(&self, type_id: TypeId) -> Option<&Type> {
-        self.types.get(type_id.0)
+    // apply the `A U Nothing => A` rule
+    if *rhs == Type::Nothing {
+        return Ok(assign_to);
     }
+
+    let rvalue_typing = ura.get_types(rvalue.reef).unwrap().typing;
+
+    if let Some(implicit) = rvalue_typing.implicits.get(&rvalue.type_id) {
+        let implicit =
+            get_type(*implicit, ura).unwrap_or_else(|| panic!("cannot find type {implicit:?}`"));
+        if lhs == implicit {
+            return Ok(assign_to);
+        }
+    }
+    Err(UnificationError())
 }
 
-pub const ERROR: TypeId = TypeId(0);
-pub const NOTHING: TypeId = TypeId(1);
-pub const UNIT: TypeId = TypeId(2);
-pub const BOOL: TypeId = TypeId(3);
-pub const EXIT_CODE: TypeId = TypeId(4);
-pub const INT: TypeId = TypeId(5);
-pub const FLOAT: TypeId = TypeId(6);
-pub const STRING: TypeId = TypeId(7);
+/// Unifies multiple type identifiers in any direction.
+pub fn convert_many<I: IntoIterator<Item = TypeRef>>(
+    ura: &UniversalReefAccessor,
+    types: I,
+) -> Result<TypeRef, UnificationError> {
+    let mut types = types
+        .into_iter()
+        .filter(|ty| ty.is_ok() && !ty.is_nothing());
+
+    let first = types.next().unwrap_or(NOTHING);
+    types.try_fold(first, |acc, ty| {
+        convert_description(ura, acc, ty).or_else(|_| convert_description(ura, ty, acc))
+    })
+}
+
+pub fn get_type<'a>(type_ref: TypeRef, ura: &'a UniversalReefAccessor) -> Option<&'a Type> {
+    ura.get_types(type_ref.reef)
+        .and_then(|types| types.typing.get_type(type_ref.type_id))
+}
+
+/// Finds the type reference from an annotation.
+pub(crate) fn resolve_type(
+    ura: &UniversalReefAccessor,
+    reef_id: ReefId,
+    env_id: SourceId,
+    type_annotation: &ast::r#type::Type,
+) -> TypeRef {
+    match type_annotation {
+        ast::r#type::Type::Parametrized(param) => {
+            if param.path.len() > 1 || !param.params.is_empty() {
+                unimplemented!();
+            }
+            if let InclusionPathItem::Symbol(_, _) = param
+                .path
+                .first()
+                .expect("Type annotation should not be empty")
```

Comment:
Any reason to perform this check since the symbol is not used as it is already resolved?

Reply:
no it's just an oversight